### PR TITLE
ALB access restriction 

### PR DESCRIPTION
**Pull request by odise**
By separating the ingress for port 80/443 to a resource it is possible to overwrite the CIDR with an extension or via the `templates` section in `mu.yml`. 

Along the way I defined a dependency between the access log property of the ALB and the bucket access policy.

### CODE REVIEW

File: templates/assets/cloudformation/elb.yml [hunk 1]
```diff
@@ -135,7 +135,7 @@ Resources:
               AWS:
                 Fn::Sub:
                   - "arn:${AWS::Partition}:iam::${ElbAccountId}:root"
-                  - ElbAccountId: !FindInMap [Regions, !Ref "AWS::Region", ElbAccountId]
+                  - ElbAccountId: !FindInMap [Regions, Ref: "AWS::Region", ElbAccountId]
```
Comment (jeffb4) [inline templates/assets/cloudformation/elb.yml:138]: Was this change needed, particularly?

File: templates/assets/cloudformation/elb.yml [hunk 2]
```diff
@@ -210,21 +219,14 @@ Resources:
       Tags:
       - Key: Name
         Value: !Ref AWS::StackName
+      - !If [ShouldCreateElbAccessLogsS3Bucket, {"Key": 'mu:bucketpolicy', "Value": {"Ref": "ElbAccessLogsS3BucketPolicy"}}, !Ref 'AWS::NoValue']
```
Comment (jeffb4) [inline templates/assets/cloudformation/elb.yml]: Stylistically can you have this conditional match the higher line count format? It is easier to discern what is going on.

File: templates/assets/cloudformation/elb.yml [hunk 3]
```diff
@@ -210,21 +219,14 @@ Resources:
...
-            - Fn::Select:
-                - 0
-                - [!Ref ElbAccessLogsS3Bucket, !Ref ElbAccessLogsS3BucketPolicy]
-            - !Ref ElbAccessLogsS3BucketName
-      - Key: access_logs.s3.prefix
-        Value: !Ref ElbAccessLogsS3Prefix
+      - Key: 'access_logs.s3.enabled'
+        Value: !If [ShouldCreateElbAccessLogsS3Bucket, 'true', 'false']
+      - !If [ShouldCreateElbAccessLogsS3Bucket, {"Key": 'access_logs.s3.prefix', "Value": {"Ref": "ElbAccessLogsS3Prefix"}}, !Ref 'AWS::NoValue']
+      - !If [ShouldCreateElbAccessLogsS3Bucket, {"Key": 'access_logs.s3.bucket', "Value": {"Ref": "ElbAccessLogsS3Bucket"}}, !Ref 'AWS::NoValue']
```
Comment (jeffb4) [inline templates/assets/cloudformation/elb.yml]: Stylistically can you have this conditional match the higher line count format? It is easier to discern what is going on.

File: templates/assets/cloudformation/elb.yml [hunk 4]
```diff
@@ -210,21 +219,14 @@ Resources:
...
-            - ShouldCreateElbAccessLogsS3Bucket
-            - Fn::Select:
-                - 0
-                - [!Ref ElbAccessLogsS3Bucket, !Ref ElbAccessLogsS3BucketPolicy]
-            - !Ref ElbAccessLogsS3BucketName
-      - Key: access_logs.s3.prefix
-        Value: !Ref ElbAccessLogsS3Prefix
+      - Key: 'access_logs.s3.enabled'
+        Value: !If [ShouldCreateElbAccessLogsS3Bucket, 'true', 'false']
+      - !If [ShouldCreateElbAccessLogsS3Bucket, {"Key": 'access_logs.s3.prefix', "Value": {"Ref": "ElbAccessLogsS3Prefix"}}, !Ref 'AWS::NoValue']
```
Comment (jeffb4) [inline templates/assets/cloudformation/elb.yml]: Stylistically can you have this conditional match the higher line count format? It is easier to discern what is going on.

**Comment by jeffb4** [comment]
Reviewed. There are some stylistic changes I'd ask you to make. Your code is most likely valid, it's just a difficult read and will be more difficult to maintain. Thanks!

File: templates/assets/cloudformation/elb.yml (hunk 1 above)
Comment (odise) [reply templates/assets/cloudformation/elb.yml:138]: > Reviewed. There are some stylistic changes I'd ask you to make. Your code is most likely valid, it's just a difficult read and will be more difficult to maintain. Thanks!

Actually the original syntax gave me a hard time. As the template is parsed and merged with other snippets coming as extensions it turned out to be the only working solution. I'm also not a fan of mixing yaml and json syntax but hose are nested constructs.

Comment (odise) [reply templates/assets/cloudformation/elb.yml:138]: > Was this change needed, particularly?

`- ElbAccountId: !FindInMap [Regions, !Ref "AWS::Region", ElbAccountId]`

Here is what I get:

`18:29:58.354 aws ▶ 061   Create stack complete err=ValidationError: Template error: Unable to get mapping for Regions::AWS::Region::ElbAccountId`

Comment (mneil) [reply templates/assets/cloudformation/elb.yml:138]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-findinmap.html

> You can't nest two instances of two functions in short form.

```
Fn::FindInMap : [Regions, !Ref "AWS::Region", ElbAccountId]
```

^ would be preferred to use.

I'm honestly surprised that `ElbAccountId: !FindInMap [Regions, Ref: "AWS::Region", ElbAccountId]` might work as well since "Ref:" is not an intrinsic function AFAIK.

Comment (odise) [reply templates/assets/cloudformation/elb.yml:138]: > https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-findinmap.html
> 
> > You can't nest two instances of two functions in short form.
> 
> ```
> Fn::FindInMap : [Regions, !Ref "AWS::Region", ElbAccountId]
> ```
> ^ would be preferred to use.
> 
> I'm honestly surprised that `ElbAccountId: !FindInMap [Regions, Ref: "AWS::Region", ElbAccountId]` might work as well since "Ref:" is not an intrinsic function AFAIK.

Let me answer with this ;):
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ref.html

Ultimately both versions are correct. However the template that finally will be deployed is generated. So you need to count in the yaml parser. I'm not familiar with your codebase but the error I mentioned earlier in this conversation indicates that the rendered result is not functioning with `!Ref`.

Comment (odise) [reply templates/assets/cloudformation/elb.yml:138]: How about merging the PR and releasing 1.5.11 :) ?

Comment (jeffb4) [reply templates/assets/cloudformation/elb.yml:138]: Resolving this part of the PR review, as there is a good reason (YAML syntax limitation) to use "Ref:" instead of "!Ref" in the array.

Leaving this PR open so Michael can investigate whether the later conditionals have to be in that form or whether any problems using the more verbose form can be resolved.

File: templates/assets/cloudformation/elb.yml (hunk 2 above)
Comment (jeffb4) [reply templates/assets/cloudformation/elb.yml]: PR author's comment on this at https://github.com/stelligent/mu/pull/428#discussion_r272642746

**Comment by jeffb4** [comment]
LGTM

**Review by jeffb4 — APPROVED** [review]
Looks good. Needs an integration test run by us.

File: templates/assets/cloudformation/elb.yml (hunk 1 above)
Comment (mneil) [reply templates/assets/cloudformation/elb.yml:138]: ugh, I had a typo in my original comment. I meant to type out the long form of Ref. Anyway...

Thanks for the link to the docs on Ref. I've corrected that several times as error not realizing it's valid without the preceding !

**Comment by odise** [comment]
Thank you for merging the PR. Any plans to (pre)release 1.5.11 soon?

**Comment by mneil** [comment]
I will work on getting out the next release this week. I'm putting it out in the public so I have to do it!
